pkg/liquidity-source/curve/twocrypto-ng/pool_simulator.go

```diff
@@ -0,0 +1,213 @@
+)
+
+type PoolSimulator struct {
+	pool.Pool
+
+	precisionMultipliers []uint256.Int
+	Reserves             []uint256.Int // same as pool.Reserves but use uint256.Int
+
+	NotAdjusted bool
+
+	Extra       Extra
+	StaticExtra StaticExtra
+
+	gas int64
+}
+
+func NewPoolSimulator(entityPool entity.Pool) (*PoolSimulator, error) {
+	sim := &PoolSimulator{}
+	if err := json.Unmarshal([]byte(entityPool.StaticExtra), &sim.StaticExtra); err != nil {
+		return nil, err
+	}
+
+	if err := json.Unmarshal([]byte(entityPool.Extra), &sim.Extra); err != nil {
+		return nil, err
+	}
+
+	var numTokens = len(entityPool.Tokens)
+	if numTokens != NumTokens {
+		return nil, ErrInvalidNumToken
+	}
+
+	if entityPool.Reserves == nil || len(entityPool.Reserves) != numTokens {
+		return nil, ErrInvalidReserve
+	}
+
+	var tokens = make([]string, numTokens)
+	var reservesBI = make([]*big.Int, numTokens)
+
+	sim.Reserves = make([]uint256.Int, numTokens)
+	sim.precisionMultipliers = make([]uint256.Int, numTokens)
+
+	for i := 0; i < numTokens; i += 1 {
+		tokens[i] = entityPool.Tokens[i].Address
+
+		reservesBI[i] = bignumber.NewBig10(entityPool.Reserves[i])
+		if err := sim.Reserves[i].SetFromDecimal(entityPool.Reserves[i]); err != nil {
+			return nil, err
+		}
+
+		sim.precisionMultipliers[i].Exp(
+			number.Number_10,
+			uint256.NewInt(uint64(18-entityPool.Tokens[i].Decimals)),
+		)
+	}
+
+	sim.Pool = pool.Pool{
+		Info: pool.PoolInfo{
+			Address:    strings.ToLower(entityPool.Address),
+			ReserveUsd: entityPool.ReserveUsd,
+			SwapFee:    bignumber.ZeroBI,
+			Exchange:   entityPool.Exchange,
+			Type:       entityPool.Type,
+			Tokens:     tokens,
+			Reserves:   reservesBI,
+			Checked:    false,
+		},
+	}
+
+	sim.gas = DefaultGas
+
+	return sim, nil
+}
+var zz = map[string]bool{}
+func (t *PoolSimulator) CalcAmountOut(param pool.CalcAmountOutParams) (*pool.CalcAmountOutResult, error) {
+	tokenAmountIn := param.TokenAmountIn
+	tokenOut := param.TokenOut
+	// swap from token to token
+	var tokenIndexFrom = t.Info.GetTokenIndex(tokenAmountIn.Token)
+	var tokenIndexTo = t.Info.GetTokenIndex(tokenOut)
+	if tokenIndexFrom >= 0 && tokenIndexTo >= 0 {
+		var amountOut, fee, amount uint256.Int
+		amount.SetFromBig(tokenAmountIn.Amount)
+
+		swapInfo := SwapInfo{}
+		err := t.GetDy(
+			tokenIndexFrom,
+			tokenIndexTo,
+			&amount,
+			&amountOut, &fee, &swapInfo.K0, swapInfo.Xp[:],
+		)
+		if err != nil {
+			return &pool.CalcAmountOutResult{}, err
+		}
+
+		if amountOut.IsZero() {
+			return &pool.CalcAmountOutResult{}, ErrZero
+		}
+
+		return &pool.CalcAmountOutResult{
+			TokenAmountOut: &pool.TokenAmount{
+				Token:  tokenOut,
+				Amount: amountOut.ToBig(),
+			},
+			Fee: &pool.TokenAmount{
+				Token:  tokenOut,
+				Amount: fee.ToBig(),
+			},
+			Gas:      t.gas,
+			SwapInfo: swapInfo,
+		}, nil
+	}
+	return &pool.CalcAmountOutResult{}, fmt.Errorf("tokenIndexFrom %v or tokenIndexTo %v is not correct", tokenIndexFrom, tokenIndexTo)
```

Comment:
I think we can rewrite the code like this

```suggestion
	if tokenIndexFrom < 0 || tokenIndexTo < 0 {
		return &pool.CalcAmountOutResult{}, fmt.Errorf("tokenIndexFrom %v or tokenIndexTo %v is not correct", tokenIndexFrom, tokenIndexTo)
	}

	var amountOut, fee, amount uint256.Int
	amount.SetFromBig(tokenAmountIn.Amount)

	swapInfo := SwapInfo{}
	err := t.GetDy(
		tokenIndexFrom,
		tokenIndexTo,
		&amount,
		&amountOut, &fee, &swapInfo.K0, swapInfo.Xp[:],
	)
	if err != nil {
		return &pool.CalcAmountOutResult{}, err
	}

	if amountOut.IsZero() {
		return &pool.CalcAmountOutResult{}, ErrZero
	}

	return &pool.CalcAmountOutResult{
		TokenAmountOut: &pool.TokenAmount{
			Token:  tokenOut,
			Amount: amountOut.ToBig(),
		},
		Fee: &pool.TokenAmount{
			Token:  tokenOut,
			Amount: fee.ToBig(),
		},
		Gas:      t.gas,
		SwapInfo: swapInfo,
	}, nil
```

Reply:
copied from tricrypto, let me update for both